src/rf/js/src/home/components/modals.js

```diff
@@ -107,9 +188,224 @@ var UploadModal = React.createBackboneClass({
+        }
+
+        if (!this.state.description) {
+            errors.description = 'Please enter a description.';
+        }
+
+        function setDateErrors(state, errors, fieldName) {
+            if (!state[fieldName]) {
+                errors[fieldName] = 'Please enter a valid date.';
+            } else {
+                var date = moment(state[fieldName]);
+                if (date.year() < 1800) {
+                    errors[fieldName] = 'Please enter a year after 1500';
+                } else if (date.year() > 2500) {
+                    errors[fieldName] = 'Please enter a year before 2500';
+                }
+            }
+        }
+
+        setDateErrors(this.state, errors, 'capture_start');
+        setDateErrors(this.state, errors, 'capture_end');
+
+        if(this.state.capture_start && this.state.capture_end) {
+            var start = moment(this.state.capture_start),
+                end = moment(this.state.capture_end);
+            if (end.isBefore(start)) {
+                errors.capture_dates = 'Please enter a capture end that is not before the start.';
+            }
+        }
+
+        if (!this.state.area) {
+            errors.area = 'Please enter an area.';
+        }
+
+        return errors;
+    },
+
+    getServerErrors: function(responseJSON) {
+        var errors = this.getInitialServerErrors();
+
+        if (responseJSON.errors.duplicate_name) {
+            errors.duplicate_name = 'Please choose a name not already in use.';
+        }
+
+        return errors;
+    },
+
+    getUpdatedServerErrors: function(oldState) {
+        // serverErrors should be removed if the input
+        // value that led to the error has been changed.
+        // Of course, the new value could still lead to a server error later.
+        var errors = this.state.serverErrors;
+        if (oldState.name != this.state.name) {
+            errors.duplicate_name = '';
+        }
+        return errors;
+    },
+
+    goToPane: function(paneInd) {
+        this.setState({activePane: paneInd});
+    },
+
+    backPane2: function(e) {
+        e.preventDefault();
+        this.goToPane(1);
+    },
+
+    backPane3: function(e) {
+        e.preventDefault();
+        this.goToPane(2);
+    },
+
+    updateState: function(callback) {
+        // Update the state using values in input elements,
+        // update client and server errors, and then call
+        // optional callback.
+        var self = this,
+            oldState = this.state;
+
+        self.setState(self.getLayerFields(), function() {
+            self.setState({clientErrors: self.getClientErrors()}, function() {
+                self.setState({serverErrors: self.getUpdatedServerErrors(oldState)},
+                              callback || _.noop);
+            });
+        });
+    },
+
+    attemptContinuePane1: function(e) {
+        e.preventDefault();
+        var self = this;
+        this.updateState(function() {
```

Comment:
It seems like you should be able to validate the form before updating the state. That could eliminate this step. Ex.

```js
if (self.pane1Valid()) {
    self.goToPane(2);
}
// If there were client-side errors, the state is updated, and the component is re-rendered.
// No need to do anything else.
```